Use current file directory in File Rename dialog

mere-human opened the pull request:
Fix #10095

donho commented:
It doesn't work.
Have you tested the PR you provided? :)

mere-human commented:
> It doesn't work, and it makes even a regression.
> Have you tested the PR you provided? :)

Yes, I tested it on Windows 10 and Windows 7 VM.
Does rename still uses the Default Directory for you?

mere-human commented:
How to test:
* Open Notepad++
* Menu > Settings > Preferences > Default Directory
* Change to `%USERPROFILE%\Desktop`
* Open some text file (not in Desktop folder)
* File > Rename
* Expected: Save As dialog is opened in the file's directory

mere-human commented:
Here is a video of the new behavior:

https://user-images.githubusercontent.com/9664141/124386482-582d9000-dce3-11eb-9345-9cbe182a01e1.mp4



Note the directory after Rename command is different from the directory after Open command.

sasumner commented:
I tested it as well.
With _Follow current document_ in effect, the directory that is opened is very strange.  (Note that I did not try the other two choices for _Default Open/Save Directory_.)
It is one that I frequently use in my daily-use version of N++, but that directory is not known to my running debug instance.
It is certainly not the folder of the current document.
(On Windows 10.)

donho commented:
@mere-human 
Here's the result of my test of Rename op:

![image](https://user-images.githubusercontent.com/90293/124386770-3d0b5200-dcdc-11eb-8441-cdf60e4b8999.png)

donho commented:
BTW, can "Save as..." to be changed into "Rename..." on title bar?

mere-human commented:
> It is certainly not the folder of the current document.

> Here's the result of my test of Rename op:

Ok, thank you all for testing it. I will look into the problems you've mentioned.


> BTW, can "Save as..." to be changed into "Rename..." on title bar?

I think it is possible. I can try.

mere-human commented:
I fixed all the problems and updated **How to test** [above](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/10100#issuecomment-873587168).

donho commented:
It doesn't work when `Remember used directory` is set.
The way to reproduce it:
Save a doc as c:\temp\a.txt, choose a file with different path, say c:\foo\b.txt
then do"Rename" & "Save as" several times without click OK (but cancel 
then go to c:\faa\c.txt and "Save as" - the last used directory becomes c:\foo\

BTW, the caption should not be "Rename current tab" but just "Rename" is OK.

mere-human commented:
> then go to c:\faa\c.txt and "Save as" - the last used directory becomes c:\foo\

Isn't it the expected behavior? "Save as" follows the last used directory. "Rename" does not. In the previous step, you used "Save as" for `c:\foo\b.txt`:

> then do"Rename" & "Save as" several times without click OK (but cancel

 so it was the last used directory displayed for `c:\faa\c.txt`.

donho commented:
> Isn't it the expected behavior? 

No. It should always "c:\temp\" since "c:\temp\a.txt," was the last saved file. On `b.txt` & `c.txt` the "save as" dialog is only evoked but not confirmed (ie. OK button pushed down). Therefore the paths of these 2 files should not be considered as last used directories.

> "Save as" follows the last used directory

Not if the Cancel button in Save As dialog is clicked.

mere-human commented:
> It should always be "c:\temp" since "c:\temp\a.txt," was the last saved file. On `b.txt` & `c.txt` the "save as" dialog is only evoked but not confirmed (ie. OK button pushed down). Therefore the paths of these 2 files should not be considered as last used directories.

This is a different problem.
I can reproduce it starting from Notepad++ version 8.0 (portable). It is not present in version 7.9.5.
I would suggest to create a new issue and fix it separately. What do you think?

mere-human commented:
Ok, I actually don't want to wait for response or create a new issue myself.
Moreover, the fix is 1 line of change.
So I'll fix it here rightaway.

mere-human commented:
I updated How to test comment [above](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/10100#issuecomment-873587168).
Also, I found out that the behavior of _Remember last used directory_ is not properly documented: https://npp-user-manual.org/docs/preferences/#default-directory

donho commented:
@mere-human 
>Also, I found out that the behavior of Remember last used directory is not properly documented: https://npp-user-manual.org/docs/preferences/#default-directory

Which part?
But for me, the current behaviour is not logical. I created an issue for that:
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/10115

mere-human commented:
> Which part?

I described it in this issue https://github.com/notepad-plus-plus/npp-usermanual/issues/258
Basically, the part about changing the current directory only if the dialog was confirmed is missing.

> But for me, the current behaviour is not logical. I created an issue for that:

Yes, I agree. This is a regression in v8.0.
